### PR TITLE
Channel.dev

### PR DESCRIPTION
**Pull request by wenweil**
created channel endpoints

channel creation endpoints moved to /server and /user respectively
RUD endpoints created
some early tests for the RUD endpoints

slight name change in message controller
one new error message

### CODE REVIEW

File: src/controllers/channel.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            const err = new Error(errorCodes.CHANNEL_NOT_EXIST);
+            err.name = errorCodes.CHANNEL_NOT_EXIST;
+            throw err;
+          }
+          res.json(channel);
+        } catch (error) {
+          next(error);
+        }
+      },
+      deleteChannel: async (req, res, next) => {
```
Comment (wenhao-zhang) [inline src/controllers/channel.js:32]: I think you need to check if the user is allowed to delete the channel first. You probably also need to delete the channel from the associated server, which you can add when we have it set up, or from the users list of DMs.

File: src/controllers/channel.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+          const deleted = await channelRepository.delete(opt);
+
+          res.json(deleted);
+        } catch (error) {
+          next(error);
+        }
+      },
+      editChannel: async (req, res, next) => {
+        try {
```
Comment (wenhao-zhang) [inline src/controllers/channel.js:59]: Same as above, you probably want to check if the user is allowed to do this action before it is done.

File: src/controllers/message.js [hunk 3]
```diff
@@ -46,14 +46,14 @@ module.exports = [
       getRoom: async (req, res, next) => {
```
Comment (wenhao-zhang) [inline src/controllers/message.js]: Why do you have two ways to do the same thing?

File: src/repositories/channel.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
+const { get } = require("lodash")
+
+module.exports = [
+    "channelModel",
+    (channelModel) => {
+        //todo : finish this
+        const get = async (opt) => {
+            const channel = channelModel.findOne(opt.query);
```
Comment (wenhao-zhang) [inline src/repositories/channel.js]: .lean(opts.lean) here to use lean optionally

File: src/utils/channeltypes.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
+module.exports = [
+    () => {
```
Comment (wenhao-zhang) [inline src/utils/channeltypes.js]: Can you put this into values as a property instead?

**Review by wenhao-zhang — CHANGES_REQUESTED** [review]

File: src/controllers/message.js (hunk 3 above)
Comment (wenweil) [reply src/controllers/message.js]: artifact of the past, still useful for some manual testing

Comment (wenweil) [reply src/controllers/message.js]: will be removing though

File: src/controllers/channel.js (hunk 1 above)
Comment (wenweil) [reply src/controllers/channel.js:32]: will implement once the channel create endpoints in user and server are finished

Comment (wenhao-zhang) [reply src/controllers/channel.js:32]: Okay that should be fine then. I'll make you an issue as a reminder as well

File: tests/Unit/controllers/channel.test.js [hunk 6]
```diff
@@ -0,0 +1,257 @@
+const errorsFactory = require("../../../src/utils/error/errors");
```
Comment (wenhao-zhang) [inline tests/Unit/controllers/channel.test.js:1]: Can you add some unit test for catching general errors. Just make sure that any errors get handled properly

File: src/repositories/channel.js [hunk 7]
```diff
@@ -7,7 +7,7 @@ module.exports = [
         const get = async (opt) => {
             const channel;
             if(opt.hasOwnProperty('lean')){
-                channel = channelModel.findOne(opt.query).lean;
+                channel = channelModel.findOne(opt.query).lean();
```
Comment (wenhao-zhang) [inline src/repositories/channel.js]: you can just pass in the value to it. You don't even have to check. If the value is false or undefined it just won't run the query as lean.

Comment (wenweil) [reply src/repositories/channel.js]: okay

Comment (wenweil) [reply src/repositories/channel.js]: well, feels like its more work than necessary to change every query function so it always contains a lean property

File: tests/Unit/controllers/channel.test.js (hunk 6 above)
Comment (wenweil) [reply tests/Unit/controllers/channel.test.js:1]: not too sure what u mean by that

File: src/repositories/channel.js (hunk 7 above)
Comment (wenhao-zhang) [reply src/repositories/channel.js]: But you don't need to add lean to every query function? You only need to add it to the get and not every call to get you need to set it to lean.

Comment (wenhao-zhang) [reply src/repositories/channel.js]: I don't think you have to set that many queries to lean and its not any less work then attaching lean the options object every time if you are referring to calling the get function.

File: src/repositories/channel.js [hunk 8]
```diff
@@ -7,7 +7,7 @@ module.exports = [
         const get = async (opt) => {
             const channel;
             if(opt.hasOwnProperty('lean')){
```
Comment (wenhao-zhang) [inline src/repositories/channel.js]:
```suggestion
           const channel = channelModel.findOne(opt.query).lean(opts.lean);      
```

You already setting it to true. If the value is false or not set, the query will not be executed as a lean.

File: tests/Unit/controllers/channel.test.js (hunk 6 above)
Comment (wenhao-zhang) [reply tests/Unit/controllers/channel.test.js:1]: I would like to see a test where you cause a random error and see that the code catches it properly. The purpose of this is to ensure there is always error catching.

Comment (wenhao-zhang) [reply tests/Unit/controllers/channel.test.js:1]: From discussion, we determined this test is not necessary

**Review by wenhao-zhang — COMMENTED** [review]
LGTM

**Review by wenhao-zhang — APPROVED** [review]

**Pull request merged**